mansrc/getSeq.1
'\" t
.\"     Title: getseq
.\"    Author: [see the "AUTHORS" section]
.\" Generator: Asciidoctor 1.5.5
.\"      Date: 
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "GETSEQ" "1" "" "\ \&" "\ \&"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\\$2 \(laURL: \\$1 \(ra\\$3
..
.if \n[.g] .mso www.tmac
.LINKSTYLE blue R < >
.SH "NAME"
getseq \- print genome sequences from databases on the standard output
.SH "SYNOPSIS"
.sp
MySQL:
.RS 4
\fBgetSeq\fP [OPTIONS] \-\-species=\fISPECIES\fP \-\-seq=\fISEQUENCE\fP \-\-dbaccess=\fIdbname,host,user,passwd\fP
.RE
.sp
SQLite:
.RS 4
\fBgetSeq\fP [OPTIONS] \-\-species=\fISPECIES\fP \-\-seq=\fISEQUENCE\fP \-\-dbaccess=\fIdbname.db\fP \-\-speciesfilenames=\fISPECIESFILENAMES\fP
.RE
.SH "DESCRIPTION"
.sp
Retrieve a genome sequence from a MySQL and SQLite database or a part of it.
Optionally the reverse complement is printed. Print to standard output.
.SH "OPTIONS"
.SS "Mandatory options"
.sp
\fB\-s\fP, \fB\-\-species\fP=\fISPECIES\fP
.RS 4
SPECIES is the species identifier used when loading the sequence into the database.
.RE
.sp
\fB\-q\fP, \fB\-\-seq\fP=\fISEQUENCE\fP
.RS 4
SEQUENCE is the ID of the sequence to retrieve.
.RE
.sp
\fB\-c\fP, \fB\-\-dbaccess\fP=\fIdatabase\fP
.RS 4
.sp
MySQL:
.RS 4
a string containing dbname, host, user, passwd joined with a comma
.sp
dbname
.RS 4
the name of the SQL database
.RE
.sp
host
.RS 4
the host name or IP
.RE
.sp
user
.RS 4
the database user
.RE
.sp
passwd
.RS 4
users and password
.RE
.RE
.sp
SQLite:
.RS 4
the file name of the SQLite database (must contain a dot)
.RE
.RE
.sp
\fB\-f\fP, \fB\-\-speciesfilenames=SPECIESFILENAMES\fP
.RS 4
This parameter is only recognized for SQLite databases.
.br
SPECIESFILENAMES is the file where the species identifier and the related file names of the sequences are stored.
.br
Format:
.RE
.sp
.if n \{\
.RS 4
.\}
.nf
    Homo sapiens <TAB> /dir/to/genome/human.fa
    Mus musculus <TAB> /dir/to/genome/mouse.fa
.fi
.if n \{\
.RE
.\}
.SS "Additional options"
.sp
\fB\-a\fP, \fB\-\-start\fP=\fInumber\fP
.RS 4
Retrieve subsequence starting at this position (coordinates are 1\-based).
.RE
.sp
\fB\-b\fP, \fB\-\-end\fP=\fInumber\fP
.RS 4
Retrieve subsequence ending at this position(coordinates are 1\-based).
.RE
.sp
\fB\-r\fP, \fB\-\-rc\fP
.RS 4
Output the reverse complement of the sequence.
.RE
.sp
\fB\-h\fP, \fB\-\-help\fP
.RS 4
Produce help message.
.RE
.SH "EXAMPLE"
.sp
.if n \{\
.RS 4
.\}
.nf
  MySQL:
    getSeq \-\-species=mouse \-\-seq=chr16 \-\-dbaccess=vertebrates,localhost,myuser,mypasswd
    getSeq \-\-species=human \-\-seq=chr21 \-\-start=32026 \-\-end=32085 \-\-rc \-\-dbaccess=vertebrates,localhost,myuser,mypasswd

  SQLite:
    getSeq \-\-species=mouse \-\-seq=chr16 \-\-dbaccess=vertebrates.db \-\-speciesfilenames=speciesfilenames.tsv
.fi
.if n \{\
.RE
.\}
.SH "AUTHORS"
.sp
AUGUSTUS was written by M. Stanke, O. Keller, S. König, L. Gerischer and L. Romoth.

HISTORY.TXT
     - remove bam2wig
     - compile with CGP and ZIPINPUT support by default
     - compile with SQLite and MySQL support for CGP (switch of by adding SQLite = false and MySQL = false to common.mk)
List of changes from version 3.3.2 to 3.3.3 (until Sep 13th, 2019)
     - docker container
     - new script tfix_in_frame_stop_codon_genes.py that replaces genes where spliced in-frame stop codons were predicted
     - new scripts: compare_masking.pl, merge_masking.pl
     - new program bamToWig.py as alternarive to bam2wig C++ binary
     - fix warnings on new GCC compiler (8.3)
     - introduced unit tests of C++ code (make test)
     - bugfixes (lex.cc, autoAug.pl)
     - DIAMOND as alternative to BLAST in aa2nonred.pl
List of changes from version 3.3.1 to 3.3.2 (until Oct 5th, 2018)
     - bugfixes in comparative augustus, utrrnaseq
     - new species Chiloscyllium punctatum (bamboo shark), Scyliorhinus torazame (cat shark), Rhincodon typus (whale shark)
     - updated comparative augustus (CGP) tutorial 
List of changes from version 3.3 to 3.3.1 (until May 8th, 2018)
     - new species pisaster (Pisaster ochraceus, ochre starfish)
     - bugfixes of sampling error in intron model and in check of
       frame-compatibility of CDSpart hints without frame
     - new auxiliary program utrrnaseq to find training UTRs
     - new script setStopCodonFreqs.pl
List of changes from version 3.2.3 to 3.3 (until July 11th, 2017)
     - new program ESPOCA to estimate selective pressure on codon alignments
     - gene finding on ancestral genomes is enabled
     - new default parameters for comparative gene prediction (CGP)
     - clade parameters training for CGP
     - compatibility to Ian Fiddes' Comparative Annotation Toolkit (CAT)
     - new scripts eval_dualdecomp.pl,
     - more tolerant tree parsing
     - bugfixes in augustus, joingenes, load2sqlitedb, transMap2hints.pl, splitMfasta.pl, intron2exex.pl,
       aln2wig
     - new functionality in homGeneMapping, joingenes
     - new species ciona, mnemiopsis_leidyi, nematostella_vectensis, strongylocentrotus_purpuratus
List of changes from version 3.2.2 to 3.2.3 (until December 6th, 2016)
     - bugfixes in joingenes, load2sqlitedb, transMap2hints.pl
     - added script gtf2bed.pl
     - CGP parameters for clade of 8 primates (cgp/log_reg_parameters_primates.cfg)
     - added option --clean to load2sqlitedb
     - now filter out truncated coding genes with CDS length <= 3
     - Sascha Steinbiss: remove dependency on PATH_MAX
List of changes from version 3.2.1 to 3.2.2 (until March 30th, 2016)
     - added extrinsic.M.RM.PB.cfg for PacBio CCS hints with GMAP
       (see http://bioinf.uni-greifswald.de/bioinf/wiki/pmwiki.php?n=Augustus.PacBioGMAP )
     - standardization of Makefile
     - improvements in auxiliary program joingenes
     - added missing program load2sqlitedb
     - comparative gene predictions (CGP):
         - hints with source=M (e.g. existing annotations) are fixed
	 - new standard parameters for exon_gain, exon_loss and others
     - bugfixes, in particular a double free or corruption bug in PPX that occured with BUSCO
     - new species Volvox carteri (volvox)
List of changes from version 3.2 to 3.2.1 (until October 23rd, 2015)
     - comparative gene prediction (CGP): make the parameters from the GCB proceedings paper match again
     - new CGP default parameters (tested on 5 clades)
     - CGP version penalizes very long introns equivalently to a geometric distribution
     - code now under artistic license only
List of changes from version 3.1 to 3.2 (until September 25th, 2015)
     - comparative gene prediction (CGP option) has sensible default parameters of the logistic regression model
     - new tool hal2maf_split.pl for splitting an alignment for parallel comparative prediction
     - bugfixes to augustus (compiler warnings on MaxOS)
     - improvements to prediction of noncoding genes
     - homGeneMapping can determine homologous sets of transcripts
     - bugfixes in joingenes and bam2hints (rare segmentation fault)
     - made load2sqlitedb more robust, new option to skip index building when loading many species
     - new tutorial that includes iterative training
     - new option to shift coordinates in hint input when sequence window is cut out
     - adapted cegma2gff.pl to cegma version 2.5
List of changes from version 3.0.3 to 3.1 (until April 29th, 2015)
     - prediction of noncoding genes (ncRNA) with option --nc=on, requires input of hints, e.g. from RNA-Seq,
       predicts coding genes (with UTRs) and noncoding genes simultaneously
     - speed improvement with --UTR=1 (by 83% on human without sampling, by 18% on fly with sampling)
     - softmasking support: --softmasking=1 is a little more accurate and much faster on human
     - small overall accuracy improvement
     - prevent the prediction of spliced in-frame stop codons (for SHORT or HINTED introns only)
     - new species: rice, Danio rerio (zebrafish), s_aureus (Staphylococcus aureus),
       camponotus_floridanus (ant) contributed by Shishir K Gupta,
       sulfolobus_solfataricus (archaeon), Parasteatoda tepidariorum (house spider), Apis dorsata (from Francisco Camara Ferreira)
     - update to parameters of Thermoanaerobacter tencongensis, E.coli, Yarrowia lipolytica, schistosoma2 (Schistosoma mansoni)
     - remove dependency on libboost-filesystems, boost is now completely optional in single-species mode
     - new auxiliary program 'homGeneMapping' for identification of homologous genes in a set of gene predictions
     - new auxiliary program 'joingenes' for combining gene sets
     - bugfixes in load2db, length distribution skew of bacteria, comparative mode (findGeneRanges,
       Alignment::mergeable, memory leak), some ass hints were previously not used although possible
     - decreased minintronlen default of bam2hints to 32 as the old threshold tended to be too large
     - computing of omega = dN/dS more efficiently

src/getSeq.cc
/*
 * getSeq.cc
 *
 * License: Artistic License, see file LICENSE.TXT or
 *          https://opensource.org/licenses/artistic-license-1.0
 *
 * Description: retrieve genomic sequences from a mysql database
 */

// Project includes
#include "randseqaccess.hh"

// standard C/C++ includes
#include <getopt.h>     /* for getopt_long; standard getopt is in unistd.h */
#include <stdlib.h>     /* for exit() */

using namespace std;

int fold=60; // line width of output sequence

void printUsage();
void printSeq(string sequence,int length);
/*
 * main
 */
int main( int argc, char* argv[] ){

    int intMax = std::numeric_limits<int>::max();

    int c;
    string species, seqname;
    int start = 1;
    int end = intMax;
    Strand strand=plusstrand;
    int help=0;

    static struct option long_options[] = {
	{"dbaccess",1, 0, 'c'},
        {"species",1, 0, 's'},
	{"speciesfilenames", 1, 0,'f'},
        {"seq",1, 0, 'q'},
        {"start",1, 0, 'a'},
        {"end",1, 0, 'b'},
	{"rc",0,0,'r'},
	{"help",0,0,'h'},
        {0,0,0,0}
    };
    int option_index = 0;
    while ((c = getopt_long(argc, argv, "c:s:f:q:a:b:rh", long_options, &option_index)) != -1) {
        switch(c)
            {
	    case 'c':
		Constant::dbaccess = optarg;
                break;
            case 's':
                species = optarg;
                break;
	    case 'f':
		Constant::speciesfilenames = optarg;
		break;
            case 'q':
                seqname = optarg;
                break;
            case 'a':
                start = atoi(optarg);
                break;
            case 'b':
                end = atoi(optarg);
                break;
            case 'r':
                strand=minusstrand;
                break;
	    case 'h':
		help=1;
		break;
	    }
    }

    if (help){
	printUsage();
	exit(1);
    }
    if (species.empty()){
	cerr << "Missing species name. Required parameter." << endl;
	printUsage();
	exit(1);
    }
    if (seqname.empty()){
	cerr << "Missing sequence name. Required parameter." << endl;
	printUsage();
	exit(1);
    }

    if (Constant::dbaccess.empty()){
	cerr << "Missing database access info. dbaccess is a required parameter." << endl;
	printUsage();
	exit(1);
    }

    if(end < start || start < 1 || end < 1){
	cerr << "Not a genomic interval. Typo in the start or end coordinate?" << endl;
	exit(1);
    }

    DbSeqAccess *rsa=NULL;

    if (Constant::dbaccess.find(',') != string::npos){ // assuming mysql access
	cerr << "assuming a MySQL database" << endl;
#ifdef M_MYSQL
	rsa = new MysqlAccess;
#else
	cerr << "Database access not possible with this compiled version. Please recompile with flag MYSQL = true added to common.mk." << endl;
	exit(1);
#endif
	
    }
    else if(Constant::dbaccess.find('.') != string::npos){ // assuming sqlite access
	cerr << "assuming a SQLite database" << endl;
	if(Constant::speciesfilenames.empty()){
	    cerr << "Missing parameter speciesfilenames." << endl;
	    exit(1);
	}
#ifdef M_SQLITE
	rsa = new SQLiteAccess(Constant::dbaccess.c_str());
#else
	cerr <<"Database access not possible with this compiled version. Please recompile with flag SQLITE = true set in common.mk." << endl;
	exit(1);
#endif
    }

    try{
	AnnoSequence *annoseq = rsa->getSeq(species, seqname, start-1, end-1, strand);
	if(annoseq){
	    if(annoseq->offset + annoseq->length < end ){
		if(end < intMax){
		    cerr <<"Warning: end position "<< end <<" is past the end of the sequence." << endl;
		}
		end = annoseq->offset + annoseq->length;
		cerr <<"Retrieving "<< seqname << ":" << start << "-" << end << endl;
	    }
	    // print fasta header
	    cout << ">"<< seqname << " " << start << " " << end;
	    if(strand == plusstrand)
		cout << " +" <<endl;
	    else
		cout << " -" <<endl;
	    // print sequence
	    printSeq(annoseq->sequence, end-start+1);
	}
    }
    catch(ProjectError &e){
	cerr << "random sequence access failed on " << species << ", " << seqname << ":"
	     << start << "-" << end << endl;;
	cerr << e.getMessage() << endl;
	exit(1);
    }
    exit(1);
}

void printUsage(){
    cerr << "usage:\n\
MySQL:\n\
    getSeq [parameters] --species=SPECIES --seq=SEQUENCE --dbaccess=dbname,host,user,passwd \n\
SQLite:\n\
    getSeq [parameters] --species=SPECIES --seq=SEQUENCE --dbaccess=dbname.db --speciesfilenames=SPECIESFILENAMES\n\
\n\
SPECIES is the species identifier used when loading the sequence into the database\n\
SEQUENCE is the ID of the sequence to retrieve\n\
dbname,host,user,passwd are the name of the SQL database, the host name or IP, the database user and password\n\
SPECIESFILENAMES is the file where the species identifier and the related file names of the sequences are stored \n\
    format: Homo sapiens <TAB> /dir/to/genome/human.fa\n\
            Mus musculus <TAB> /dir/to/genome/mouse.fa\n\
\n\
parameters:\n\
--help        print this usage info\n\
--rc          output the reverse complement of the sequence\n\
--start=N     retrieve subsequence starting at position N (coordinates are 1-based)\n\
--end=N       retrieve subsequence ending at position N (coordinates are 1-based)\n\
\n\
example MySQL:\n\
     getSeq --species=hg19 --seq=chr21 --dbaccess=saeuger,localhost,myuser,mypasswd \n\
     getSeq --species=hg19 --seq=chr21 --start=47870612  --end=48086047 --rc --dbaccess=saeuger,localhost,myuser,mypasswd \n\
\n\
example SQLite:\n\
    getSeq --species=hg19 --seq=chr21 --start=47870612 --end=48086047 --dbaccess=genomes.db --speciesfilenames=genomes.tbl\n\
\n";
}

void printSeq(string sequence,int length){
    int start=0, end;
    while(start < length){
	end = (start + fold < length)? (start + fold - 1) : length - 1;
	cout << sequence.substr(start,end-start+1) << endl;
	start+=fold;
    }
}
